urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/KeyGenerator.java

```diff
@@ -0,0 +1,178 @@
+
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+import java.util.Vector;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+
+import javax.xml.bind.DatatypeConverter;
+
+import sun.security.util.ObjectIdentifier;
+import sun.security.x509.AlgorithmId;
+import sun.security.x509.AuthorityKeyIdentifierExtension;
+import sun.security.x509.BasicConstraintsExtension;
+import sun.security.x509.CertificateAlgorithmId;
+import sun.security.x509.CertificateExtensions;
+import sun.security.x509.CertificateIssuerName;
+import sun.security.x509.CertificateSerialNumber;
+import sun.security.x509.CertificateSubjectName;
+import sun.security.x509.CertificateValidity;
+import sun.security.x509.CertificateVersion;
+import sun.security.x509.CertificateX509Key;
+import sun.security.x509.ExtendedKeyUsageExtension;
+import sun.security.x509.KeyIdentifier;
+import sun.security.x509.SubjectKeyIdentifierExtension;
+import sun.security.x509.X500Name;
+import sun.security.x509.X509CertImpl;
+import sun.security.x509.X509CertInfo;
+
+/**
+ * Generate RSA keys and X.509 self-signed certificates for SSL/TLS
+ * 
+ * @author Guillermo Grandes / guillermo.grandes[at]gmail.com
+ */
+@SuppressWarnings("restriction")
+public class KeyGenerator {
+	private static final String KEYPAIR_ALG = "RSA";
+	private static final String SIGNATURE_ALG = "SHA256withRSA";
+
+	public static void main(String[] args) throws Exception {
+		final String iam = KeyGenerator.class.getName();
+		if (args.length < 4) {
+			System.out.println("java " + iam + " <bits> <days> <CommonName> <filename-without-extension>");
+			System.out.println("");
+			System.out.println("Example:");
+			System.out.println("  java " + iam + " 2048 365 TestServer test");
+			System.out.println("");
+			System.out.println("* Output files are named <filename>.crt & <filename>.key");
+			return;
+		}
+		int i = 0;
+		final int bits = Integer.parseInt(args[i++]);
+		final int days = Integer.parseInt(args[i++]);
+		final String cn = args[i++];
+		final String file = args[i++];
+		//
+		// Generate RSA Key
+		final KeyPairGenerator kgAsym = KeyPairGenerator.getInstance(KEYPAIR_ALG);
+		kgAsym.initialize(bits);
+		final KeyPair kp = kgAsym.genKeyPair();
+		File keyFile = new File(file + ".key");
+		writeKey(new FileOutputStream(keyFile), kp.getPrivate());
+		// Clear permissions
+		keyFile.setExecutable(false, false);
+		keyFile.setReadable(false, false);
+		keyFile.setReadable(true);
+		keyFile.setWritable(false, false);
+		keyFile.setWritable(true);
+		//
+		// Generate & SelfSign X.509
+		final X509Certificate crt = generateCertificate("CN=" + cn, kp, days, SIGNATURE_ALG);
+		final File crtFile = new File(file + ".crt");
+		writeCertificate(new FileOutputStream(crtFile), crt);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the resource created by the following constructor: *FileOutputStream*. The resource is referenced at line 93. The resource is closed at line 164. There are other execution paths that do not contain closure statements, for example, when KeyGenerator.writeBufferBase64 (in method writeCertificate) throws an exception. Ensure the object returned by *FileOutputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).